### PR TITLE
fix: Ensure patching fails gracefully

### PR DESCRIPTION
**Pull request by tssweeney**
This PR improves our resilience to patching errors. It correctly catches all patching errors and logs the issue only once to the user.

This also introduces a test suite to:
1. Assert that we correctly catch these issues (and sets up testing for next pr)
2. Makes sure that all gobbled errors are thrown in test suites

### CODE REVIEW

File: weave/trace_server/async_batch_processor.py [hunk 1]
```diff
@@ -64,6 +65,8 @@ def _process_batches(self) -> None:
                     if e.response.status_code == 413:
                         # 413: payload too large, don't raise just log
                         logger.error(f"Error processing batch: {e}")
+                        if get_raise_on_captured_errors():
```
Comment (gtarpenning) [inline weave/trace_server/async_batch_processor.py:68]: nice, was going to suggest that

File: weave/trace/op_extensions/error_once.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from typing import Callable
+
+LOG_ONCE_MESSAGE_SUFFIX = " (subsequent messages of this type will be suppressed)"
+logged_messages = []
+
+
+def log_once(log_method: Callable[[str], None], message: str) -> None:
```
Comment (gtarpenning) [inline weave/trace/op_extensions/error_once.py:7]: a bit confusing wrt the file name `error_once`. Also a docstring here would be nice (with the cannonical example)

Comment (gtarpenning) [reply weave/trace/op_extensions/error_once.py:7]: this is to be used mainly in generators? for loops?

**Review by gtarpenning — APPROVED** [review]
We could really use codecov! I know @jamie-rasmussen  mentioned it but it would be especially useful for these kinds of changes (broad but shallow).

File: weave/trace/op_extensions/log_once.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+from typing import Callable
+
+LOG_ONCE_MESSAGE_SUFFIX = " (subsequent messages of this type will be suppressed)"
+logged_messages = []
+
+
+def log_once(log_method: Callable[[str], None], message: str) -> None:
```
Comment (andrewtruong) [inline weave/trace/op_extensions/log_once.py:7]: Is the intent to use this with many different loggers?

If just python logging, in future you can use the builtin `addFilter`

**Review by andrewtruong — APPROVED** [review]

**Pull request merged**